hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/replication/ECUnderReplicationHandler.java

```diff
@@ -0,0 +1,242 @@
+     // State is UNDER_REPLICATED
+      final List<DatanodeDetails> deletionInFlight = new ArrayList<>();
+      for (ContainerReplicaOp op : pendingOps) {
+        if (op.getOpType() == ContainerReplicaOp.PendingOpType.DELETE) {
+          deletionInFlight.add(op.getTarget());
+        }
+      }
+      List<Integer> missingIndexes = replicaCount.unavailableIndexes(true);
+      // We got the missing indexes, this is excluded any decommissioning
+      // indexes. Find the good source nodes.
+      if (missingIndexes.size() > 0 && !replicaCount.unRecoverable()) {
+        List<ContainerReplica> sources = replicas.stream().filter(r -> r
+            .getState() == StorageContainerDatanodeProtocolProtos
+            .ContainerReplicaProto.State.CLOSED)
+            // Exclude stale and dead nodes. This is particularly important for
+            // maintenance nodes, as the replicas will remain present in the
+            // container manager, even when they go dead.
+            .filter(r -> ReplicationManager
+                .getNodeStatus(r.getDatanodeDetails(), nodeManager).isHealthy())
+            .filter(r -> !deletionInFlight.contains(r.getDatanodeDetails()))
+            .collect(Collectors.toList());
+        LOG.debug("Missing indexes detected for the container {}." +
+                " The missing indexes are {}", id, missingIndexes);
+        // We have source nodes.
+        if (sources.size() >= repConfig.getData()) {
+          final List<DatanodeDetails> selectedDatanodes =
+              getTargetDatanodes(replicas, container, missingIndexes.size());
+
+          List<ReconstructECContainersCommand.DatanodeDetailsAndReplicaIndex>
+              sourceDatanodesWithIndex = new ArrayList<>();
+          for (ContainerReplica srcReplica : sources) {
+            sourceDatanodesWithIndex.add(
+                new ReconstructECContainersCommand
+                    .DatanodeDetailsAndReplicaIndex(
+                    srcReplica.getDatanodeDetails(),
+                    srcReplica.getReplicaIndex()));
+          }
+
+          final ReconstructECContainersCommand reconstructionCommand =
+              new ReconstructECContainersCommand(id.getProtobuf().getId(),
+                  sourceDatanodesWithIndex, selectedDatanodes,
+                  int2byte(missingIndexes),
+                  repConfig);
+          // Keeping the first target node as coordinator.
+          return ImmutableMap.of(selectedDatanodes.get(0),
+              reconstructionCommand);
+        } else {
+          LOG.warn("Cannot proceed for EC container reconstruction for {}, due"
+              + "to insufficient source replicas found. Number of source "
+              + "replicas needed: {}. Number of available source replicas are:"
+              + " {}. Available sources are: {}", container.containerID(),
+              repConfig.getData(), sources.size(), sources);
+        }
+      }
+
+      if (containerHealthResult.underReplicatedDueToDecommission()) {
```

Comment:
This isn't correct, from the java doc:

```
    /**
     * Indicates whether the under-replication is caused only by replicas
     * being decommissioned or entering maintenance. Ie, there are not replicas
     * unavailable.
     * @return True is the under-replication is caused by decommission.
     */
    public boolean underReplicatedDueToDecommission() {
      return dueToDecommission;
    }
```
This only returns true when the under-replication is only due to decommission. If there are also unavailable indexes, it returns false. You probably need to get the count of `decommissionOnlyReplicas().size() > 0`

Reply:
oh now I remember why I put if else.
Since we removed if else, every time we check decom, this may not meet when missingIndexes>0 already. There if else make sense.
Since we don't need to do this in if else, we need decom only counter here. I will fix this.

Reply:
used decomOnly indexes check.